### PR TITLE
Introduces Permissions to the core.

### PR DESCRIPTION
**Pull request by Vassyli**
Adds a permission model, a manager, as well as traits and interfaces required for models who want to implement permissions.

Every actor has many permission associations, every permission association references a permission as well as a state. This results in ternary permission system:

- Allowed (actor has a permission association for this permission id, with state = Allowed)
- Denied (actor has a permission association for this permission id, with state = Denied)
- Undefined/Not set (actor has not a permission association for this permission id).

This is later important if a crate wants to implement a multi-layered permission system with groups and inheritance.

### CODE REVIEW

File: src/Models/PermissionableInterface.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?php
+declare(strict_types=1);
+
+namespace LotGD\Core\Models;
+
+/**
+ * Implement this interface if an entity has associates permissions.
```
Comment (austenmc) [inline src/Models/PermissionableInterface.php]: nit: "has associated permissions"

**Comment by austenmc** [comment]
Good observation about supporting inherited permissions.

File: src/PermissionManager.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * isDenied           | False | False   | True
+ * hasPermissionSet   | False | True    | True
+ */
+class PermissionManager
+{
+    const Allowed = 1;
+    const Denied = -1;
+    
+    const Superuser = "lotgd/core/superuser";
+    const AddScenes = "lotgd/core/scene/add";
```
Comment (austenmc) [inline src/PermissionManager.php:40]: I think these scene/character-based permissions should be added as part of the scene or character class.

File: tests/ModelTestCase.php [hunk 3]
```diff
@@ -47,6 +47,7 @@
                 $libraryConfigurationManager = new LibraryConfigurationManager($composerManager, getcwd());
                 $directories = $libraryConfigurationManager->getEntityDirectories();
                 $directories[] = implode(DIRECTORY_SEPARATOR, [__DIR__, '..', 'src', 'Models']);
+                $directories[] = implode(DIRECTORY_SEPARATOR, [__DIR__, 'Ressources', 'TestModels']);
```
Comment (austenmc) [inline tests/ModelTestCase.php]: nit: "Resources"

File: src/Tools/Model/Permissionable.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    protected function loadPermissions()
+    {
+        if (empty($this->_permissions)) {
+            foreach ($this->permissions as $permission) {
+                $this->_permissions[$permission->getId()] = $permission;
+            }
+        }
+    }
+
+    public function hasPermission(string $permissionId): bool
```
Comment (austenmc) [inline src/Tools/Model/Permissionable.php]: `hasPermissionSet` is clearer that it's a presence flag. `hasPermission` sounds like "Do I have permission?" which sounds like "Is allowed?".

**Comment by austenmc** [comment]
I'm concerned there is some unnecessary complexity here that will cause problems for the noob coders we expect to be writing permission checks.
1. In the current PR, there are two ways to check for a permission: one is through the PermissionManager and one is through the actor class. It's hard to know, just from reading the APIs of User and PermissionManager, which one to use.
2. Presumably, the cause of (1) is b/c you want to have permissionable entities that aren't Users, like maybe Groups? Are there any other entities you think will implement `PermissionableInterface`? I can imagine maybe modules have permissions, if you want to establish stronger boundaries between the code you pull in off the web. I'm not sure how useful this is honestly, or whether it's worth the complexity. What if we restricted permissions to be only for Users, then you check permissions via the Manager, and the manager implements a hook that can support Groups via a module.

Open to your thoughts here.

**Comment by Vassyli** [comment]
Yes, that's what I intended. Now that you mention it, there is not really much anything besides User and Groups that might use this. I didn't even think about modules. Maybe bots, but they could easily just be a User instead.

I will refactor the Permissionable trait to an abstract Actor model class instead. Also, I will provide a PermissionGroup Interface that the PermissionManager can understand and work with it, as well as add events at least to "isAllowed" and "isDenied".

Alternatively, instead of events I could introduce a GroupableInterface that a User entity might or might not implement whose only function returns a PermissionGroup. Or additionally to the events.

Thoughts?

**Comment by austenmc** [comment]
Still planning on working on this?

As far as your question is concerned, I would still with something simple, just the events, for now. We're losing steam at this point :) so I say we push toward some kind of MVP.

**Comment by Vassyli** [comment]
Yes, I'm still planning, just didn't find time to sit down and get it right.

File: src/Models/Actor.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+use LotGD\Core\Models\PermissionAssociationInterface;
+
+/**
+ * This abtract class provides functionality for user entities that crates might
+ * want to implement, such as permissions.
+ */
+abstract class Actor
+{
+    /** @var array Associations between permission-id and PermissionAssociation entity. */
+    private $_permissions = [];
```
Comment (austenmc) [inline src/Models/Actor.php]: nit: is prefixing private vars with _ in a style guide somewhere? Is that common in PHP?

File: src/PermissionManager.php [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+namespace LotGD\Core;
+
+use LotGD\Core\Exceptions\PermissionIdNotFoundException;
+use LotGD\Core\Models\Actor;
+use LotGD\Core\Models\PermissionableInterface;
+use LotGD\Core\Models\Permission;
+
+/**
+ * The PermissionManager manages (checks and manipulates) of actors.
```
Comment (austenmc) [inline src/PermissionManager.php]: nit: "manipulates) permissions of actors"

File: src/PermissionManager.php [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+ *
+ * The PermissionManager class provides methods to work with permissions and is
+ * the only way to check and manipulate permissions. It can be used to create or
+ * delete permissions, to remove, allow or deny permissions to actors and to
+ * check whether an actor has a certain permission or if it is explicitely
```
Comment (austenmc) [inline src/PermissionManager.php]: nit: explicitly
nit: denied *to* him

File: tests/Resources/TestModels/User.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+declare(strict_types=1);
+
+namespace LotGD\Core\Tests\Ressources\TestModels;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\Table;
+
+use Lotgd\Core\Models\Actor;
+#use LotGD\Core\Models\PermissionableInterface;
```
Comment (austenmc) [inline tests/Resources/TestModels/User.php]: Remove the commented out code?

**Review by austenmc — APPROVED** [review]

File: src/Models/Actor.php (hunk 5 above)
Comment (Vassyli) [reply src/Models/Actor.php]: No it's not. I need to get rid of it, that's something that carried over from python - thanks for pointing it out.

**Pull request merged**